app/controllers/users/sessions_controller.rb

```diff
@@ -44,19 +55,42 @@ def alive?
       current_user.present? && session_alive
     end
 
-    def track_authentication_attempt(email)
-      user = User.find_by(email: email.downcase) || AnonymousUser.new
+    def existing_user
+      @_existing_user ||= User.find_by(email: params[:user][:email].downcase) || AnonymousUser.new
+    end
 
+    def track_authentication_attempt
       properties = {
-        success?: current_user.present?, user_id: user.uuid
+        success?: current_user.present?, user_id: existing_user.uuid
       }
 
       analytics.track_event(Analytics::EMAIL_AND_PASSWORD_AUTH, properties)
     end
 
+    def check_concurrent_sessions
+      return unless existing_user.is_a? User
+      return unless existing_user.decorate.too_many_sessions?
+
+      flash[:error] = t('errors.messages.concurrent_sessions')
+      redirect_to root_url
+    end
+
     def cache_active_profile
       cacher = Pii::Cacher.new(current_user, user_session)
       cacher.save(params[:user][:password])
     end
+
+    def track_user_session
+      session_syncer.clean(current_user)
+      IdentityLinker.new(current_user, Identity::LOCAL, session.id).link_identity
+    end
+
+    def session_syncer
+      SessionSyncer.new(session_store)
+    end
+
+    def session_store
+      session.instance_variable_get('@by')
```

Comment:
Hmm. This raises my eyebrows as well. It seems strange that we'd have to be tied to the session store implementation to do this. I would expect it to work with any session store. Have you considered using Warden to do this, like the [devise_security_extension](https://github.com/phatworx/devise_security_extension/blob/b2ee978af7d49f0fb0e7271c6ac074dfb4d39353/lib/devise_security_extension/hooks/session_limitable.rb) gem does?